blog/2021-03-16-BatchSolvingWithAmq.adoc

```diff
@@ -0,0 +1,182 @@
+* A teacher can teach at most one lesson at the same time (hard).
+* A student can attend at most one lesson at the same time (hard).
+* A teacher prefers to teach in a single room (soft).
+* A teacher prefers to teach sequential lessons and dislikes gaps between lessons (soft).
+* A student dislikes sequential lessons on the same subject (soft).
+
+== Quickstart structure
+
+image::projectStructure.png[Project structure]
+
+The project consists of three modules:
+
+* *amq-quarkus-school-timetabling-common* defines the problem domain, and the `SolverRequest` and `SolverResponse` classes for messaging.
+The following two modules depend on this one.
+* *amq-quarkus-school-timetabling-client* is the Client Quarkus application that contains a UI, a REST endpoint and a demo data generator.
+* *amq-quarkus-school-timetabling-solver* is the Solver Server Quarkus application that solves school timetabling problem instances coming via a message queue `solver_request`.
+
+== Messaging
+
+image::architecture.png[Architecture]
+
+The Client application serializes an unsolved `TimeTable` wrapped by the `SolverRequest` class into a JSON and sends it to the `solver_request` queue.
+The Solver Server receives the request from this queue, deserializes it and solves the `TimeTable` via OptaPlanner.
+After the solving finishes, the Solver Server wraps the `TimeTable` by the `SolverResponse` class, serializes it to a JSON and sends it to the `solver_response` queue.
+
+=== Requirements
+
+. No solver request message must be lost, even if the Solver Server crashes.
+. Any error that occurs in the Solver Server must be propagated back to the Client.
+. Invalid solver request message is sent to a dead letter queue.
+
+=== AMQ vs. Kafka
+
+One of the technologies being talked about these days is Kafka.
+It’s been used in https://medium.com/@rinu.gour123/apache-kafka-use-cases-and-applications-7a5f379b8451[various use cases], including message queues, replacing the traditional messaging systems.
+However, as much as Kafka might shine in its most typical use, delivering huge amounts of small messages, this is not the problem we are trying to solve in this post.
+There will not be millions of input problems to solve every second, but odds are problems will not be small.
+
+Depending on the scale of the input problem, the message may easily grow to tenths or even hundreds of megabytes.
+Although Kafka can be configured to support such a non-typical use case, it might fail to keep up to expectations on its more typical use cases.
+As a result, users would have to maintain a separate Kafka cluster just for this specific task.
+Better approach is to https://dzone.com/articles/processing-large-messages-with-apache-kafka[pair Kafka with a storage] and make the message contain just the identifier of a record in that storage.
+The downside is obvious: the system is no longer stateless and both sending input problems and retrieving solutions takes more than a single step.
```

Comment:
I am not sure if this article needs to explain why Kafka sucks for this use case, for 2 paragraphs.
This is a negative message. That can rub people who like kafka the wrong way.

To explain IntelliJ why is better than Eclipe, just point out all the good things about Intellij that Eclipse doesn't have answer too.

How about we make it positive message?
Section title: instead of "amq vs kafka" -> "AMQ is a natural fit"
Then 1 or 2 paragraphs to point out that AMQ can handle the problem sizes, that the input frequence isn't too big for AMQ, that transactional queues can stay open long enough to avoid state, etc.

Wdyt?